include/dr/mp/algorithms/count.hpp

```diff
@@ -0,0 +1,159 @@
+                       dr::__detail::direct_iterator(rng::begin(r)),
+                       dr::__detail::direct_iterator(rng::end(r)), pred);
+}
+
+inline auto dpl_count_if(rng::forward_range auto &&r, auto &&pred) {
+  using count_type = rng::range_difference_t<decltype(r)>;
+
+#ifdef SYCL_LANGUAGE_VERSION
+  if (rng::empty(r)) {
+    return count_type{};
+  }
+
+  return std::count_if(mp::dpl_policy(),
+                       dr::__detail::direct_iterator(rng::begin(r)),
+                       dr::__detail::direct_iterator(rng::end(r)), pred);
+#else
+  assert(false);
+  return count_type{};
+#endif
+}
+
+template <dr::distributed_range DR>
+auto count_if(std::size_t root, bool root_provided, DR &&dr, auto &&pred) {
+  using count_type = rng::range_difference_t<decltype(dr)>;
+  auto comm = mp::default_comm();
+
+  if (rng::empty(dr)) {
+    return count_type{};
+  }
+
+  if (aligned(dr)) {
+    dr::drlog.debug("Parallel count\n");
+
+    // Count within the local segments
+    auto count = [=](auto &&r) {
+      assert(rng::size(r) > 0);
+      if (mp::use_sycl()) {
+        dr::drlog.debug("  with DPL\n");
+        return dpl_count_if(r, pred);
+      } else {
+        dr::drlog.debug("  with CPU\n");
+        return std_count_if(r, pred);
+      }
+    };
+
+    auto locals = rng::views::transform(local_segments(dr), count);
+    auto local = add_counts(locals);
+
+    std::vector<count_type> all(comm.size());
+    if (root_provided) {
+      // Everyone gathers to root, only root adds up the counts
+      comm.gather(local, std::span{all}, root);
+      if (root == comm.rank()) {
+        return add_counts(all);
+      } else {
+        return count_type{};
+      }
+    } else {
+      // Everyone gathers and everyone adds up the counts
+      comm.all_gather(local, all);
+      return add_counts(all);
+    }
+  } else {
+    dr::drlog.debug("Serial count\n");
+    count_type result{};
+    if (!root_provided || root == comm.rank()) {
+      result = add_counts(dr);
+    }
+    mp::barrier();
+    return result;
+  }
+}
+
+} // namespace dr::mp::__detail
+
+namespace dr::mp {
+
+//
+// Ranges
+//
+
+// range, elem, w/wo root
+
+template <typename T, dr::distributed_range DR>
+auto count(std::size_t root, DR &&dr, const T &value) {
+  auto pred = [=](auto &&v) { return v == value; };
+  return __detail::count_if(root, true, dr, pred);
+}
+
+template <typename T, dr::distributed_range DR>
+auto count(DR &&dr, const T &value) {
+  auto pred = [=](auto &&v) { return v == value; };
+  return __detail::count_if(0, false, dr, pred);
+}
+
+// range, predicate, w/wo root
+
+template <dr::distributed_range DR>
+auto count_if(std::size_t root, DR &&dr, auto &&pred) {
+  return __detail::count_if(root, true, dr, pred);
+}
+
+template <dr::distributed_range DR> auto count_if(DR &&dr, auto &&pred) {
+  return __detail::count_if(0, false, dr, pred);
+}
+
+//
+// Iterators
+//
+
+// range, elem, w/wo root
+
```

Comment:
Should be implemented as _niebloid_ - compare implementation on [std::ranges::count, std::ranges::count_if](https://en.cppreference.com/w/cpp/algorithm/ranges/count), or implementations within dr, eg. mp::views::enumerate

Reply:
Can you please elaborate on this a bit more? Although I read up a bit on niebloids (as I've never encountered them before), I still don't really get what you mean in this context. What exactly should here be implemented as a niebloid, and more importantly why?